### PR TITLE
Added ability to specify custom escape characters for LIKE

### PR DESCRIPTION
**Pull request by JCOR11599**
There are new test cases for escape syntax.

New parameter for Like Expression.

### CODE REVIEW

File: enginetest/queries.go [hunk 1]
```diff
@@ -1750,6 +1750,25 @@ var QueryTests = []QueryTest{
 		Query:    `SELECT SUBSTRING_INDEX(mytable.s, "d", 1) AS s FROM mytable INNER JOIN othertable ON (SUBSTRING_INDEX(mytable.s, "d", 1) = SUBSTRING_INDEX(othertable.s2, "d", 1)) GROUP BY 1 HAVING s = 'secon'`,
 		Expected: []sql.Row{{"secon"}},
 	},
+
+	{
+		Query:    `SELECT * FROM specialtable t WHERE t.name LIKE "%a_%" ESCAPE 'a'`,
+		Expected: []sql.Row{sql.Row{"first_row"}, sql.Row{"second_row"}, sql.Row{"third_row"}},
+	},
+	/*
```
Comment (zachmu) [inline enginetest/queries.go:1758]: Uncomment

File: enginetest/testdata.go [hunk 2]
```diff
@@ -78,6 +78,30 @@ func createSubsetTestData(t *testing.T, harness Harness, includedTables []string
...
+			table, err = harness.NewTable(myDb, "specialtable", sql.Schema{
+				{Name: "name", Type: sql.MustCreateStringWithDefaults(sqltypes.VarChar, 20), Source: "specialtable"},
+			})
+
+			if err == nil {
+				InsertRows(t, NewContext(harness), mustInsertableTable(t, table),
+					sql.NewRow("first_row"),
+					sql.NewRow("second_row"),
+					sql.NewRow("third_row"),
+					sql.NewRow(`%`),
```
Comment (zachmu) [inline enginetest/testdata.go:92]: Need some test rows where there are characters other than the special characters

File: sql/expression/like.go [hunk 3]
```diff
@@ -130,7 +132,34 @@ func (l *Like) evalRight(ctx *sql.Context, row sql.Row) (*string, error) {
...
+
+	// Use normal patternToGoRegex when escape is empty string
+	if len(e.(string)) == 0 {
+		s := patternToGoRegex(v.(string))
+		return &s, nil
+	}
+
+	// e should be exactly one character
+	if len(e.(string)) > 1 {
+		return nil, sql.ErrInvalidArgument.New("ESCAPE")
```
Comment (zachmu) [inline sql/expression/like.go:159]: Need an engine test for this condition

File: sql/expression/like.go [hunk 4]
```diff
@@ -160,10 +189,61 @@ func patternToGoRegex(pattern string) string {
...
+		if escaped {
+			escaped = false
+		}
+	}
+
+	buf.WriteRune('$')
+	return buf.String()
+}
+
+func patternToGoRegexCustom(pattern, escape string) string {
```
Comment (zachmu) [inline sql/expression/like.go:220]: patternToGoRegexWithEscape

File: sql/parse/parse.go [hunk 5]
```diff
@@ -442,6 +442,7 @@ func convertShow(ctx *sql.Context, s *sqlparser.Show, query string) (sql.Node, e
 					filter = expression.NewLike(
 						expression.NewUnresolvedColumn("Table"),
 						expression.NewLiteral(s.ShowTablesOpt.Filter.Like, sql.LongText),
+						nil,
```
Comment (zachmu) [inline sql/parse/parse.go:445]: Is it possible to use ESCAPE on these LIKE expressions? If so add a TODO

File: sql/parse/parse.go [hunk 6]
```diff
@@ -2654,6 +2659,16 @@ func comparisonExprToExpression(ctx *sql.Context, c *sqlparser.ComparisonExpr) (
 		return nil, err
 	}
 
+	// TODO: is this the right way to do this?
```
Comment (zachmu) [inline sql/parse/parse.go:2662]: Looks right, you can remove the TODO

**Review by zachmu — COMMENTED** [review]
Looks pretty good, but needs a bit more test cases in engine tests

File: sql/parse/parse.go (hunk 5 above)
Comment (JCOR11599) [reply sql/parse/parse.go:445]: It does seem like it is possible to use ESCAPE for SHOW ... LIKE ... statements.
MySQL throws a Syntax Error.

File: enginetest/queries.go [hunk 7]
```diff
@@ -1750,6 +1750,74 @@ var QueryTests = []QueryTest{
...
+		Query:    `SELECT * FROM specialtable t WHERE t.name LIKE "%$_%" ESCAPE '$'`,
+		Expected: []sql.Row{sql.Row{"first_row"}, sql.Row{"second_row"}, sql.Row{"third_row"}},
+	},
+	{
+		Query:    `SELECT * FROM specialtable t WHERE t.name LIKE "first$_%" ESCAPE '$'`,
+		Expected: []sql.Row{sql.Row{"first_row"}},
+	},
+	{
+		Query:    `SELECT * FROM specialtable t WHERE t.name LIKE "%$_row" ESCAPE '$'`,
+		Expected: []sql.Row{sql.Row{"first_row"}, sql.Row{"second_row"}, sql.Row{"third_row"}},
```
Comment (zachmu) [inline enginetest/queries.go:1767]: Table should have a row like "decoy1row" that matches `like %_row` but not `like $\_row`

**Review by zachmu — APPROVED** [review]
LGTM, just one comment in test

Tests right now can't tell if you're matching literal _ and % characters, which is the point of the escapes

**Pull request merged**